app/Connect/Actions/BuildClientPatchDataAction.php

```diff
@@ -0,0 +1,247 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Connect\Actions;
+
+use App\Models\Achievement;
+use App\Models\Game;
+use App\Models\GameAchievementSet;
+use App\Models\GameHash;
+use App\Models\PlayerGame;
+use App\Models\User;
+use App\Platform\Enums\AchievementFlag;
+use Illuminate\Database\Eloquent\Collection;
+use InvalidArgumentException;
+
+// TODO more tests
+
+// OPEN QUESTION when given a GameHash, should the root Achievements & Leaderboards be returned as `null`?
```

Comment:
This is a discussion point I'm not sure how to best handle. I also referenced this in a test:

https://github.com/RetroAchievements/RAWeb/blob/75f2b68d338d3cd43d56599719e4a04e706280bb/tests/Feature/Connect/Actions/BuildClientPatchDataActionTest.php#L402-L407

Given that the user is using a modern client and we send a list of sets to that client from `r=patch`, what should the root-level data be?
* Do root-level `Achievements` and `Leaderboards` even matter?
* Should root-level `Title`, `ImageIcon`, and `ImageIconURL` always point to the core game?
* Should `RichPresencePatch` take on the RP script of the "subset game"? 

For reference, here is a full data structure of the response in the test case above:
```php
array:2 [
  "Success" => true
  "PatchData" => array:9 [
    "ID" => 2
    "Title" => "Dragon Quest III [Subset - Bonus]"
    "ImageIcon" => "/Images/000011.png"
    "RichPresencePatch" => """
      Display:\n
      Test
      """
    "ConsoleID" => 1
    "ImageIconURL" => "/media/Images/000011.png"
    "Achievements" => array:2 [ ... two achievements ... ]
    "Leaderboards" => []
    "Sets" => array:2 [
      0 => array:8 [
        "GameAchievementSetID" => 1
        "CoreGameID" => 1
        "Title" => null
        "Type" => "core"
        "ImageIcon" => "/Images/000011.png"
        "ImageIconURL" => "/media/Images/000011.png"
        "Achievements" => array:1 [ ... one achievement ... ]
        "Leaderboards" => []
      ]
      1 => array:8 [
        "GameAchievementSetID" => 5
        "CoreGameID" => 3
        "Title" => "Bonus 2"
        "Type" => "bonus"
        "ImageIcon" => "/Images/000011.png"
        "ImageIconURL" => "/media/Images/000011.png"
        "Achievements" => array:3 [ ... three achievements ... ]
        "Leaderboards" => []
      ]
    ]
  ]
] // tests/Feature/Connect/Actions/BuildClientPatchDataActionTest.php:410
```

Reply:
> Should root-level Title, ImageIcon, and ImageIconURL always point to the core game?

I think that if the client is requesting patch data by a subset hash, the top-level data should be from the core set, and the subset as a child record.

> Do root-level Achievements and Leaderboards even matter?

The [response structure](https://github.com/RetroAchievements/rcheevos/wiki/rc_api_process_fetch_game_data_response#parameters) still has root-level fields, it would make sense to return the core data at the root level.

For legacy compatibility, when requesting a subset by game_id, the subset information should be returned at the top level, and no child records should be returned.

If we're going to eliminate the top-level records, maybe we should have a new API (i.e. `patch2`) rather than trying to overload the existing API.

> Should RichPresencePatch take on the RP script of the "subset game"?

I think subsets shouldn't have their own RP scripts. This may need further discussion.

Reply:
> In other words, let's say I load "Dragon Quest III [Subset - Bonus]". Is it correct that the top-level Title, ImageIcon, ImageIconURL, Achievements, and Leaderboards should be from "Dragon Quest III", not "Dragon Quest III [Subset - Bonus]"?

Yes.

> I can imagine some developers will get frustrated if their RP scripts attached to subsets are suddenly considered invalid.

Are there non-exclusive subsets with unique RP?

Subset:
![image](https://github.com/user-attachments/assets/8c24a80f-d643-4013-950a-3d5a4f60eae2)

Core:
![image](https://github.com/user-attachments/assets/534f588a-77fb-4631-8ed5-3fb342788d89)


Subset:
![image](https://github.com/user-attachments/assets/ae2639bf-77c8-4b54-8b3e-398b05281aa5)

Core:
![image](https://github.com/user-attachments/assets/c6b257aa-e2a0-45b2-b72f-d2b7c877bf42)


Subset:
![image](https://github.com/user-attachments/assets/e444dd7d-b3f9-4640-a7ce-5081f5c0ff1e)

Core:
![image](https://github.com/user-attachments/assets/876262ee-150b-4ae4-94e6-e7f91589e92d)


Found one. Though I could argue that the core RP is better. Maybe the prefix could be merged into that?

Subset:
![image](https://github.com/user-attachments/assets/46086f9b-d1ed-4455-8d94-d7e30e27a282)

Core:
![image](https://github.com/user-attachments/assets/073bce38-0c2c-4f4f-ae31-6045bb56b568)

Reply:
Perhaps since "specialty" and "exclusive" subsets require loading unique hashes, they could also have unique RP, but generic "bonus" sets would not.

Reply:
All this logic should now be implemented in latest, with additional tests added to verify it.

* RP patch data from specialty and exclusive subset games is now prioritized when a user loads a hash associated with one of those games.
* If the user is not globally opted out of multiset, root-level data reflects the values of the core/base game.

Reply:
Thought related to RP: how are we going to maintain/create sessions for the subsets? Will `ping` be overloaded to send all game_ids that are active? Will we just run the active subset logic on the server like we do for building the patch data? 

I thin this comes down to "specialty" and "exclusive" subsets again. Do we still need "active players" for generic bonus subsets?

Do we still need subset sessions in the database for associating unlocks, or can they be tied to the core subset?

Reply:
My gut feelings are:
* `ping` should never consider more than 1 game ID as being active.
* The active game ID should probably be the same `ID` that's sent back from `patch` at the root level.

We should probably tackle this piece in a subsequent PR.

Reply:
Something else we'll need to address in a later PR: `r=startsession` needs to return unlocks for the subsets as well so the client can disable previously earned subset achievements.